### DISCUSSION TITLE
[RFC] 079 - Edge Config 配置层

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#5215] [RFCs]
## 背景

之前做 https://github.com/lobehub/lobe-chat/discussions/913 这个的时候提过，未来会需要一个服务端的配置能力。正好最近有了一个助手市场的黑/白名单需求，在这个 RFC 中先实现一版。

## 设计思路

从终局来看，这个中间配置层应该是要兼容各种部署环境的，比如：

- 在 Vercel 上部署：直接使用 Edge Config；
- 完全本地私有化部署环境：存入 DB 并用 Redis 做缓存加速；

但是目前这一期简单做，直接写 Vercel Edge Config，后续抽象中间层，并抽取 ConfigRuntime ，到时候就可以提供 Edge Config， Redis Config 这样的 provider。

## 进展

- PR: https://github.com/lobehub/lobe-chat/pull/5216

**Comment by arvinxx** [comment]
未来实现流程：

## 添加 config key

先在 `src/server/modules/EdgeConfig` 中新增一个 config key 

```ts
enum EdgeConfigKeys {
  /**
   * Assistant whitelist
   */
  AssistantWhitelist = 'assistant_whitelist',
}
```

## 实现对应的读取方法逻辑与类型

```ts
export class EdgeConfig {

  // 实现逻辑
  getAgentWhitelist = async (): Promise<string[] | undefined> => {
    return this.client.get<string[]>(EdgeConfigKeys.AssistantWhitelist);
  };

}
```

实现应该是很简单的，主要还是包一层方法，屏蔽掉  config key 最裸的一层实现，同时补充对应的类型定义。如果这里需要添加兜底字段，也可以在这里加。

## 在对应的部分添加实现

需要有一个来是否进入 edge config 的判断逻辑，目前感觉 `!!appEnv.VERCEL_EDGE_CONFIG` 这个写法不太好，等后续想清楚了再重构掉。

```ts

      const data: AgentStoreIndex = await res.json();

      // Get the assistant whitelist from Edge Config
      const edgeConfig = new EdgeConfig();

      if (!!appEnv.VERCEL_EDGE_CONFIG) {
        const assistantWhitelist = await edgeConfig.getAgentWhitelist();

        if (assistantWhitelist && assistantWhitelist?.length > 0) {
          data.agents = data.agents.filter((item) => assistantWhitelist.includes(item.identifier));
        }
      }

      return data;
```

**Comment by Matinal111** [comment]
不了解这个设计思路的实现，不过支持 Redis 这个方向对于未来一些功能的实现会有很大帮助， Vercel 环境可以使用免费的 Redis 云数据库，私有化环境直接本地部署 Redis

**Reply by arvinxx** [reply]
你可以理解成运行时配置，比如助手市场黑白名单、默认开启的模型等等
